libs/langchain/langchain/chat_models/anthropic.py

```diff
@@ -94,29 +96,64 @@ def _convert_messages_to_prompt(self, messages: List[BaseMessage]) -> str:
             text.rstrip()
         )  # trim off the trailing ' ' that might come from the "Assistant: "
 
-    def _generate(
+    def _stream(
         self,
         messages: List[BaseMessage],
         stop: Optional[List[str]] = None,
         run_manager: Optional[CallbackManagerForLLMRun] = None,
         **kwargs: Any,
-    ) -> ChatResult:
+    ) -> Iterator[ChatGenerationChunk]:
+        prompt = self._convert_messages_to_prompt(messages)
+        params: Dict[str, Any] = {"prompt": prompt, **self._default_params, **kwargs}
+        if stop:
+            params["stop_sequences"] = stop
+
+        stream_resp = self.client.completions.create(**params, stream=True)
+        for data in stream_resp:
+            delta = data.completion
+            yield ChatGenerationChunk(message=AIMessageChunk(content=delta))
+            if run_manager:
+                run_manager.on_llm_new_token(delta)
+
+    async def _astream(
+        self,
+        messages: List[BaseMessage],
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[AsyncCallbackManagerForLLMRun] = None,
+        **kwargs: Any,
+    ) -> AsyncIterator[ChatGenerationChunk]:
         prompt = self._convert_messages_to_prompt(messages)
         params: Dict[str, Any] = {"prompt": prompt, **self._default_params, **kwargs}
         if stop:
             params["stop_sequences"] = stop
 
+        stream_resp = await self.async_client.completions.create(**params, stream=True)
+        async for data in stream_resp:
+            delta = data.completion
+            yield ChatGenerationChunk(message=AIMessageChunk(content=delta))
+            if run_manager:
+                await run_manager.on_llm_new_token(delta)
+
+    def _generate(
+        self,
+        messages: List[BaseMessage],
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[CallbackManagerForLLMRun] = None,
+        **kwargs: Any,
+    ) -> ChatResult:
         if self.streaming:
             completion = ""
-            stream_resp = self.client.completions.create(**params, stream=True)
-            for data in stream_resp:
-                delta = data.completion
-                completion += delta
-                if run_manager:
-                    run_manager.on_llm_new_token(
-                        delta,
-                    )
+            for chunk in self._stream(messages, stop, run_manager, **kwargs):
+                completion += chunk.text
```

Comment:
A more efficient way to do this is…
```python
completion = ''.join(c.text for c in self._stream(messages, stop, run_manager, **kwargs))
```
It might not be the _most_ efficient.  There could be even better techniques.

The concatenation loop you wrote appears in the code about 7 times.  Depending on the length of the stream being processed, a more efficient technique could make a big difference in performance.  You may find this functional technique can improve other classic `for` loops, too.